src/redux-middleware/pushQueue.ts

```diff
@@ -139,13 +133,54 @@ const flushPushQueue = (): Thunk<Promise<void>> => async (dispatch, getState) =>
 
   if (pushQueue.length === 0) return Promise.resolve()
 
-  dispatch(flushDeletes())
-  dispatch(flushEdits())
-  dispatch(flushMoves())
+  /**
+   * Returns updated thoughtIndexUpdates using the latest state after pulling pending lexemes.
+   */
+  const getSyncedThoughtsUpdate = (batch: PushBatch) => Object.keys(batch.thoughtIndexUpdates).reduce<Index<Lexeme | null>>((acc, thoughtId) => {
+
+    const updatedState = getState()
+
+    // Return early if this particular lexeme was not pending
+    const hadPendingLexeme = batch.pendingLexemes?.[thoughtId]
+    if (!hadPendingLexeme) return acc
+
+    // Note: Assuming thoughtIndexUpdates entry with null value won't occur, as they will have associated lexeme already pulled.
+    const lexemeInBatch = batch.thoughtIndexUpdates[thoughtId] as Lexeme
+
+    const lexemeInUpdatedState = updatedState.thoughts.thoughtIndex[thoughtId]
+    /**
+     * Returns updated lexeme context without the already availble entries in the the lexemeInBatch contexts.
+     */
+    const updatedLexemeContexts = () => updatedState.thoughts.thoughtIndex[thoughtId].contexts.filter(thoughtContext => {
+      const isAlreadyInUpdate = lexemeInBatch.contexts.some(thoughtContextInner => equalArrays(thoughtContextInner.context, thoughtContext.context))
+      return !isAlreadyInUpdate
+    })
+
+    return {
+      ...acc,
+      ...lexemeInUpdatedState ? { [thoughtId]: { ...lexemeInBatch, contexts: [...lexemeInBatch.contexts, ...updatedLexemeContexts()] } } : {}
+    }
+  }, {})
+
+  // Pull all the pending lexemes that are not available in state yet, reconcile and update the thoughtIndexUpdates. Prevents local and remote lexemes getting overriden by incomplete application state (due to lazy loading).
+  // Related issue: https://github.com/cybersemics/em/issues/1074
+  const lexemeSyncedPushQueue = await Promise.all(pushQueue.map(async batch => {
+    if (!batch.pendingLexemes || Object.keys(batch.pendingLexemes).length === 0) return batch
+    await dispatch(pullLexemes(Object.values(batch.pendingLexemes!)))
```

Comment:
I think we need to collate `pendingLexemes` and then make a single `pullLexemes` call, as one `pullLexemes` for every batch is inefficient and defeats the purpose of the `pushQueue`, which is to throttle database access. The logic is already there with `mergeBatch`. You just need to add `pendingLexemes` to `mergeBatch` and then use that instead of pulling every batch separately.

If you need to have access to `lexemeSyncedPushQueue` before `mergeBatch` is called for some reason, then you may need to merge the `pendingLexemes` manually, but I still think we want to find a way to use a single `pullLexemes`.

Let me know if this doesn't make sense or if I'm missing something!

Reply:
Thanks. It makes sense.